File: neetcode-250/linked-list/141-linked-list-cycle.ts
import { ListNode } from "./listNode";

function hasCycle(head: ListNode | null): boolean {
	let slow = head,
		fast = head?.next;

	while (fast) {
		if (slow == fast) {
			return true;
		}
		slow = slow?.next ?? null;
		fast = fast.next?.next;
	}

	return false;
}

File: neetcode-250/linked-list/141-linked-list-cycle.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return None

        slow, fast = head, head.next
        while fast:
            if slow == fast:
                return True
            slow = slow.next
            fast = fast.next
            if fast:
                fast = fast.next
        return False
